lib/travis/addons/handlers/github_status.rb

```diff
@@ -10,9 +11,17 @@ class GithubStatus < Base
         EVENTS = /build:(created|started|finished|canceled|restarted)/
 
         def handle?
-          return true if installation?
-          Addons.logger.error "No GitHub OAuth tokens found for #{object.repository.slug}" unless tokens.any?
-          tokens.any?
+          if installation? && !github_status_for_apps?
+            false
+          elsif installation? && github_status_for_apps?
+            Addons.logger.info "Commit Status posted for GitHub-Apps managed repo because of repo- or owner-level feature flag"
+            true
+          elsif tokens.empty?
+            Addons.logger.error "No Commit or Check Run Status because no GitHub Apps installation or OAuth tokens found for #{object.repository.slug}"
+            false
+          else
+            true
+          end
```

Comment:
I find this a bit tricky to follow (as in, took me a while to understand), I think simpler would be:

``` ruby
def handle?
  if installation?
    return false unless github_status_for_apps?
    Addons.logger.info "Commit Status posted for GitHub-Apps managed repo because of repo- or owner-level feature flag"
  elsif tokens.empty?
    Addons.logger.error "No Commit or Check Run Status because no GitHub Apps installation or OAuth tokens found for #{object.repository.slug}"
    return false
  end

  true
end
```

Or even:

``` ruby
def handle?
  installation? ? handle_installation? : handle_legacy?
end

def handle_installation?
  return false unless github_status_for_apps?
  Addons.logger.info "Commit Status posted for GitHub-Apps managed repo because of repo- or owner-level feature flag"
  true
end

def handle_legacy?
  return true if tokens.any?
  Addons.logger.error "No Commit or Check Run Status because no GitHub Apps installation or OAuth tokens found for #{object.repository.slug}"
  false
end
```

Alternatively, I think adding comments would help.